src/libstd/sync/condvar.rs

```diff
@@ -167,6 +163,25 @@ impl Condvar {
         }
     }
 
+    /// Wait on this condition variable for a notification, timing out after a
+    /// specified duration.
+    ///
+    /// The semantics of this function are equivalent to `wait_timeout` except
+    /// that the implementation will repeatedly wait while the duration has not
+    /// passed and the provided function returns `false`.
+    #[unstable]
+    pub fn wait_timeout_with<'a, T, F>(&self,
+                                       guard: MutexGuard<'a, T>,
+                                       dur: Duration,
+                                       f: F)
+                                       -> LockResult<(MutexGuard<'a, T>, bool)>
+            where F: FnMut(LockResult<&mut MutexGuard<'a, T>>) -> bool {
```

Comment:
On second thought, this may want to take `&mut T` as using `mem::replace` to swap out the mutex being used seems like it might be a bad idea. Conceptually `&mut MutexGuard<T>` is basically the same thing as `&mut T` as well I believe.